src/polymer/expression-scanner.ts

```diff
@@ -0,0 +1,163 @@
+
+export interface Template extends parse5.ASTNode { content: parse5.ASTNode; }
+
+export function getAllDataBindingTemplates(node: parse5.ASTNode) {
+  const toplevelTemplates = dom5.queryAll(
+      node,
+      dom5.predicates.OR(isDomBindTemplate, isDomModuleTemplate)) as Template[];
+
+  const results = new Set<Template>();
+  for (const template of toplevelTemplates) {
+    results.add(template);
+
+    const innerTemplates = dom5.queryAll(
+        template.content,
+        matchInnerDataBindingTemplate,
+        [],
+        dom5.childNodesIncludeTemplate) as Template[];
+    for (const innerTemplate of innerTemplates) {
+      results.add(innerTemplate);
+    }
+  }
+
+  return results;
+}
+
+
+/**
+ * Databinding into an attribute/property.
+ */
+export class ScannedAttributeExpression implements ScannedFeature {
+  /** The element whose attribute is being assigned to.*/
+  astNode: parse5.ASTNode;
+  /** The HTML element attribute that's being assigned to.*/
+  attribute: parse5.ASTAttribute;
+  sourceRange: SourceRange;
+  warnings: Warning[] = [];
+  directionality: 'down-only'|'bidirectional';
+  expressionText: string;
+  expressionAst: estree.Program;
+
+  /**
+   * If this is a bidirectional data binding, and an event name was specified
+   * (using ::eventName syntax), this is that event name.
+   */
+  eventName: string|undefined;
+
+  constructor(
+      astNode: parse5.ASTNode, attribute: parse5.ASTAttribute,
+      sourceRange: SourceRange, directionality: 'down-only'|'bidirectional',
+      expressionText: string, eventName: string|undefined) {
+    this.astNode = astNode;
+    this.attribute = attribute;
+    this.sourceRange = sourceRange;
+    this.directionality = directionality;
+    this.expressionText = expressionText;
+    this.eventName = eventName;
+    this.expressionAst = espree.parse(
+        expressionText,
+        Object.assign({sourceType: 'script' as 'script'}, baseParseOptions));
+  }
+}
+
+/**
+ * Find and parse Polymer databinding expressions in HTML.
+ *
+ * TODO(rictic):
+ *  - parse expressions inline in strings rather than entire attributes
+ *  - parse expressions in text nodes
+ *  - allow getting source ranges inside of a databinding expression.
+ *  - surface parse errors in expressions as warnings.
+ */
+export class ExpressionScanner implements HtmlScanner {
+  async scan(
+      document: ParsedHtmlDocument, _: (visitor: HtmlVisitor) => Promise<void>):
+      Promise<ScannedAttributeExpression[]> {
+    const results: ScannedAttributeExpression[] = [];
+    const dataBindingTemplates = getAllDataBindingTemplates(document.ast);
+    for (const template of dataBindingTemplates) {
+      dom5.nodeWalkAll(template.content, (node) => {
+        if (node.attrs) {
+          for (const attr of node.attrs) {
+            if (!attr.value) {
+              continue;
+            }
+            const oneWayMatch = attr.value.match(/^\s*\[\[\s*(.*?)\s*\]\]\s*$/);
```

Comment:
In the next PR this is significantly refactored and made more efficient to match inline expressions.